### PR TITLE
Add `disable_nested_struct_naming` option

### PR DESCRIPTION
**Pull request by oblique**
With this option the following structs will be named as `foo` and `bar` instead of `foo` and `foo_bar`.

```
struct foo {
    struct bar {
    } b;
};
```

In case of an unnamed definition we build the canonical name from the inner most named definition. For example the following will generate `baz__bindgen_ty_1`:

```
struct foo {
    struct bar {
        struct baz {
            struct {
            } u;
        } z;
    } b;
};
```

This option is needed in some rare situations where user used another code generator that already mangled nested definitions.

A real life example is [asn1c] with `-fcompound-names` option.

[asn1c]: https://github.com/vlm/asn1c

### CODE REVIEW

**Comment by oblique** [comment]
@emilio r?

**Comment by emilio** [comment]
Thank you for the ping, and sorry for the horrible lag, I've missed a bunch of my bindgen notifications lately looks like :(

File: tests/headers/disable-canonical-naming.hpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// bindgen-flags: --disable-canonical-naming
+
+struct foo {
```
Comment (emilio) [inline tests/headers/disable-canonical-naming.hpp]: So I think this needs more tests with the interactions with C++ features like:

 * Namespaces.
 * Typedefs.
 * Nested C++ classes.

This surely works on the simple case, but have we figured out what's the desired behavior for all those things?

Do we expect:

```cpp
struct foo {
    typedef int bar;
};
```

To generate a `foo_bar` typedef or a `bar` typedef?

File: src/ir/item.rs [hunk 2]
```diff
@@ -858,7 +870,30 @@ impl Item {
...
+
+        let ids: Vec<_> = if ctx.options().disable_canonical_naming {
+            let mut ids = Vec::new();
+
+            // If target is anonimous we need find its first named ancestor.
+            if target.is_anon() {
+                while let Some(id) = ids_iter.next() {
+                    ids.push(id);
+
+                    if !ctx.resolve_item(id).is_anon() {
```
Comment (emilio) [inline src/ir/item.rs]: So this breaks at the first anonymous ancestor, how do we deal with nested anonymous structs? This generates `_bindgen_ty_<n>_bindgen_ty_1`, which is not how everything else is generated, what's the reasoning for that?

File: tests/headers/disable-canonical-naming.hpp (hunk 1 above)
Comment (oblique) [reply tests/headers/disable-canonical-naming.hpp]: I didn't thought of that, thanks. The rationale is: If C++ adds scoping then we should also add it.

In your example, C++ will add scoping, so the type is `foo::bar`. So we should produce `foo_bar`.
I will add more tests and fix these issues.

File: src/ir/item.rs (hunk 2 above)
Comment (oblique) [reply src/ir/item.rs]: It breaks on the first named ancestor, not anonymous.

```c++
struct foo {
    struct {
        struct {
        } inner;
    } inner;
};

struct bar {
    struct bar2 {
        struct {
        } inner;
    } inner;
};
```

So for the above we generate:

```
foo
foo__bindgen_ty_1
foo__bindgen_ty_1__bindgen_ty_1

bar
bar2
bar2__bindgen_ty_1
```

The reason I do this is to handle anonymous structs without having conflicts.

If I wasn't adding the first named struct then we would generate `__bindgen_ty_1` two times, one for anon struct under `foo` and one for anon struct under `bar2`, which will result to conflicts.

File: tests/headers/disable-canonical-naming.hpp (hunk 1 above)
Comment (oblique) [reply tests/headers/disable-canonical-naming.hpp]: Based on my above comment, maybe we should also change the name of the feature? `disable_canonical_naming` might not be the best option.

Any ideas?

File: src/ir/item.rs (hunk 2 above)
Comment (emilio) [reply src/ir/item.rs]: Yeah, fair enough. I thought we didn't have nested `__bindgen_ty_*` but we do seem to have them already, so it seems ok.

File: src/lib.rs [hunk 3]
```diff
@@ -1064,6 +1067,27 @@ impl Builder {
...
+    /// ```c
+    /// struct foo {
+    ///     struct bar {
+    ///     } b;
+    /// };
+    /// ```
+    ///
+    /// This method disables that behavior.
+    ///
+    /// Note, that this option may cause bindgen to generate duplicate names.
```
Comment (emilio) [inline src/lib.rs]: It may be worth pointing out the C compiler option that it mimics.

**Comment by oblique** [comment]
I started implementing your suggestions but I found that C++ is namespacing also C compound types. The following code creates different type names in C++ and different in C.

```c++
struct foo1 {
    struct bar1 {
    };
};


#ifdef __cplusplus
extern "C" {
#endif
    struct foo2 {
        struct bar2 {
        };
    };
#ifdef __cplusplus
}
#endif
```

C produce's the types:

```
struct foo1
struct bar1
struct foo2
struct bar2
```

And C++ produce's the types (even within `extern "C"` block):

```
struct foo1
struct foo1::bar1
struct foo2
struct foo2::bar2
```

So what you are suggesting can not be implemented. However since the problem that I'm trying to solve is for C, I'm suggesting to just rename the current option to `disable-c-nested-struct-naming` and explicitly state that it is indented for C headers.

What do you think?

**Comment by oblique** [comment]
Ready for review. I changed option name, doc comments and commit message.

Don't worry about the lag :)

**Comment by oblique** [comment]
@emilio ping

File: src/ir/item.rs [hunk 4]
```diff
@@ -804,6 +813,10 @@ impl Item {
     ///
     /// This name should be derived from the immutable state contained in the
     /// type and the parent chain, since it should be consistent.
+    ///
+    /// If `BindgenOptions::disable_nested_struct_naming` is true then returned
+    /// name is the inner most non-anonimous name plus all the anonimous base names
```
Comment (emilio) [inline src/ir/item.rs]: nit: anonymous (with `y`, not `i`)

**Comment by oblique** [comment]
Ready!

**Comment by emilio** [comment]
Oh, I was going to do that myself, but thanks! rustfmt seemed angry at one of your newlines, but all the other travis jobs were green, so I pushed it with that fixed in https://github.com/rust-lang/rust-bindgen/commit/cb8266620596222b1cd9dbe6551cc1e3e8bb7f72.

Thanks again, and sorry again for the lag!

**Pull request closed without merging**